src/main/java/org/opensearch/ad/rest/handler/ModelValidationActionHandler.java

```diff
@@ -0,0 +1,718 @@
+                logger.error("Failed to create search request for last data point", exception);
+            });
+        searchFeatureDao.getLatestDataTime(anomalyDetector, latestTimeListener);
+    }
+
+    private void getSampleRangesForValidationChecks(
+        Optional<Long> latestTime,
+        AnomalyDetector detector,
+        ActionListener<ValidateAnomalyDetectorResponse> listener,
+        Map<String, Object> topEntity
+    ) {
+        if (!latestTime.isPresent()) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.NOT_ENOUGH_HISTORICAL_DATA,
+                        DetectorValidationIssueType.GENERAL_DATA,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        } else if (latestTime.get() <= 0) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.NOT_ENOUGH_HISTORICAL_DATA,
+                        DetectorValidationIssueType.GENERAL_DATA,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        }
+        long timeRangeEnd = Math.min(Instant.now().toEpochMilli(), latestTime.get());
+        try {
+            getBucketAggregates(timeRangeEnd, listener, topEntity);
+        } catch (IOException e) {
+            listener.onFailure(new EndRunException(detector.getDetectorId(), CommonErrorMessages.INVALID_SEARCH_QUERY_MSG, e, true));
+        }
+    }
+
+    private void getBucketAggregates(
+        long latestTime,
+        ActionListener<ValidateAnomalyDetectorResponse> listener,
+        Map<String, Object> topEntity
+    ) throws IOException {
+        List<String> featureFields = ParseUtils.getFeatureFieldNames(anomalyDetector, xContentRegistry);
+        AggregationBuilder aggregation = getBucketAggregation(latestTime);
+        BoolQueryBuilder query = QueryBuilders.boolQuery().filter(anomalyDetector.getFilterQuery());
+        if (anomalyDetector.isMultientityDetector()) {
+            if (topEntity.isEmpty()) {
+                listener
+                    .onFailure(
+                        new ADValidationException(
+                            CommonErrorMessages.CATEGORY_FIELD_NO_DATA,
+                            DetectorValidationIssueType.CATEGORY,
+                            ValidationAspect.MODEL
+                        )
+                    );
+                return;
+            }
+            for (Map.Entry<String, Object> entry : topEntity.entrySet()) {
+                query.filter(QueryBuilders.termQuery(entry.getKey(), entry.getValue()));
+            }
+        }
+        if (anomalyDetector.isMultiCategoryDetector()) {
+            for (String category : anomalyDetector.getCategoryField()) {
+                query.filter(QueryBuilders.existsQuery(category));
+            }
+        }
```

Comment:
What is the purpose of this filter? Wouldn't the keys in the previous if block encapsulate this already?

Reply:
I left this from when I was just making sure category fields exists and not getting the top entity. Removing this.